webgui-new/README.md

```diff
@@ -0,0 +1,76 @@
+# New Next.js (React) based webgui for CodeCompass
+
+## Prerequisites
+
+Make sure you are using at least version 18 of [Node.js](https://nodejs.org/en/).\
+You can also download [nvm](https://github.com/nvm-sh/nvm), and run `nvm install` at the root of the project to use the specified Node version (18.13.0).
+
+## Development
+
+For easier development, first run the webserver (CodeCompass_webserver), then run the development server.\
+Since the webserver and development server are different origins, you have to use the [CORS extension](https://chrome.google.com/webstore/detail/allow-cors-access-control/lhobafahddgcelffkeicbaginigeejlf?hl=en) to resolve CORS issues.
+
+Alternatively, you could add this change to the `MainRequestHandler::begin_request_handler(struct mg_connection* conn_)` method in `mainrequesthandler.cpp`, then build the webserver again.\
+(Note: Do not commit this change, as it is unnecessary in production.)
```

Comment:
The problem with the proxy is that it only works with actual API routes in Next.js, which is not the case in this application, so for now I'm not seeing a solution for this.

I will remove the other solution from the README to modify the main request handler, because I agree it is messy. However, I think for development, using the CORS extension I linked is perfectly fine, because it is easy to use (you just have to toggle a button in the browser), and is only required for development.

Reply:
I am not that familiar with Next.JS, but I think the `http-proxy-middleware` package should not be affected by it, since it is a lower level package below the framework, modifying the middleware.

The browser extension option could be mentioned, but that is a 3rd party, Chrome-specific solution to this problem. I think the application itself should provide a solution for this, proxying in the development server is a widely used approach in my experience. (Such a solution was used in all Angular and React projects I have worked so far.)

Reply:
Next.js provides an option in the `next.config.js` file, called `rewrites`. However, this can only be used to proxy client-side routes. The `http-proxy-middleware` package you mentioned can only be used for Node.js webservers.

Unfortunately, you cannot set the request header of a webserver from a client-side application. For this, you would have to modify the webserver's request handler (like the way I mentioned earlier in the README).

Reply:
> Next.js provides an option in the `next.config.js` file, called `rewrites`. However, this can only be used to proxy client-side routes. The `http-proxy-middleware` package you mentioned can only be used for Node.js webservers.

In development mode, Next.js runs a Node.js webserver as well. Otherwise, it won't be able to serve the frontend.
That is the reason of the CORS error, since this Node.js webserver runs on a different port than the backend.
https://nextjs.org/docs/deployment#nodejs-server

We will only use the proxy in development mode. In production a static build is deployed on the same host and port as the backend, so no CORS issue will be raised.

> If I understand correctly, you cannot set the request header of a webserver from a client-side application. For this, you would have to modify the webserver's request handler (like the way I mentioned earlier in the README).

The proxy server standing between the frontend and the backend will solve this, that is the purpose of it.

Reply:
Then I misunderstood something. I will look into this, but for know I don't know how would I implement something like this in my case.

Reply:
Update:

Apparently, there is a way to override the Next.js server and use a custom Node.js server, which I tried, and it works. Here I can set the proxy, and the CORS issues are resolved.

However, I need to use the `NEXT_PUBLIC_BACKEND_URL` `.env` variable for the proxy, and only Next.js provides an option to override `.env` variables with a `.env.local` file, Node.js does not. And because the Node.js server starts before the Next.js app initializes, the overrided `.env` variables won't be available. Therefore, `.env.local` is not a valid option anymore if we want a proxy server.

Reply:
Update:

I found out that you can define a custom Node.js webserver that overrides the Next.js webserver. This way, the proxy can be created. However, there are still some issues I need to resolve.

Reply:
This was my original answer, but I replied to another thread, so I will just copy it here for future reference:

> I have used the `http-proxy-middleware` package with environment files for Node.js based proxying to avoid CORS issues in multiple projects, it works.
>
> Such an example is the [TMS](https://gitlab.com/tms-elte/frontend-react) submission management project developed at the ELTE FI in React. See `.env` files and the `src/setupProxy.js` for reference.
>
> I will get on to review the other parts of the PR. If you are stuck with how to implement this, we shall get back to this sometime later. Let's keep this thread open as a reminder.

Reply:
I've made little progess. I found out that it is possible to overwrite the Next.js development server with a custom Node.js server, where you can use the `http-proxy-middleware` package.

The proxy works, however, it proxies the contents of the whole webserver. This means that when running the development server, the contents of the proxied webserver will be visible, and not the Next.js application.

I understand that there is a way to only proxy certain requests, like `/api`, just as in the example you provided, at TMS. But in my case, there are no separated API requests. I tried filtering out all the requests that end in `Service`, because only the requests like `/WorkspaceService` and `/ProjectService` would need to be proxied, the others would need to be handled by Next.js, but so far, I didn't have any luck.

I commited these changes to a separate branch on my fork, `proxy-server`, because it think it is a good start.
https://github.com/mdeme01/CodeCompass/commit/e4b8ed7b67ac6a97ba517e4cd4471fcb1bc7ec4a

Reply:
> I've made little progess. I found out that it is possible to overwrite the Next.js development server with a custom Node.js server, where you can use the http-proxy-middleware package.

Sounds good. If we would like to use the `Next.JS` development server, there also seems to be a `next-http-proxy-middleware` package for the framework, which can add the required proxy functionality to Next.JS.

> I understand that there is a way to only proxy certain requests, like /api, just as in the example you provided, at TMS. But in my case, there are no separated API requests. I tried filtering out all the requests that end in Service, because only the requests like /WorkspaceService and /ProjectService would need to be proxied, the others would need to be handled by Next.js, but so far, I didn't have any luck.

Maybe the easiest solution for this to really serve the Thrift requests with a path prefix like `/api`. Regardless of the development environment, this logical separation of the backend and frontend can be beneficial anyway. (E.g. it enables to deploy the frontend and the backend separately later.)

Reply:
Update:

I managed to solve it. I modifed the config to use the current URL for the Thrift service at all times, not just in production. This way the `/Service` paths can be proxied too to the webserver. Then I used a `pathFilter` to only proxy the paths ending with `/Service`, the others are handled by Next.js.

It was a little unclear to me how would I create the proxy with the `next-http-proxy-middleware` package (because it is dependant on the built-in `/api` routes of Next.js), so I sticked with a custom Node.js server.

It is also worth mentioning that the `pathFilter` option is only available in the v3 beta verison of `http-proxy-middleware`, but it is really mandatory to use in this case. However, I read that the v3 will be out of beta soon.

And also another note, this way, the proxying also works with `rewrites` in the `next.config.js`. However, the requests were doubled because of the rewrites and the app was slower, so I decided not to go with this option.

Reply:
Great! I will take a look on this later, whether it could work with NextJS's own webserver, but the provided solution is fine already.

Reply:
@mdeme01 Well, I also found it hard to integrate `next-http-proxy-middleware`, so your proposed solution with a [Custom Server](https://nextjs.org/docs/pages/building-your-application/configuring/custom-server) should be fine, it is an officially documented and advised solution.